dst/dst_test.go

```diff
@@ -0,0 +1,547 @@
+	}
+
+	t.Log("Running DST using random seed:", os.Getenv(randomSeedKey))
+	logger := newTestLogger(t)
+	objectStorage := frostdb.NewDefaultObjstoreBucket(
+		objstore.NewInMemBucket(), frostdb.StorageWithLogger(log.WithPrefix(logger, "component", "objstore")),
+	)
+	storageDir := t.TempDir()
+	logStoreWrapper := &testLogStore{}
+	walTicker := &fakeTicker{c: make(chan time.Time, 1)}
+	storeID := 0
+	c, err := newStore(
+		storageDir, log.WithPrefix(logger, "storeID", storeID), objectStorage, func(logStore internalWal.LogStore) internalWal.LogStore {
+			logStoreWrapper.LogStore = logStore
+			return logStoreWrapper
+		}, walTicker,
+	)
+	require.NoError(t, err)
+	defer c.Close()
+
+	ctx := context.Background()
+	var db atomic.Pointer[frostdb.DB]
+	{
+		// Separate scope to avoid table pointer misuse.
+		newDB, err := c.DB(ctx, dbName)
+		require.NoError(t, err)
+		db.Store(newDB)
+	}
+
+	tp := &tableProvider{}
+	tableConfig := frostdb.NewTableConfig(samples.SampleDefinition())
+	{
+		// Separate scope to avoid table pointer misuse.
+		table, err := db.Load().Table(tableName, tableConfig)
+		require.NoError(t, err)
+		tp.Update(table)
+	}
+
+	t.Log("DB initialized, starting commands")
+	w := &writerHelper{logger: logger, tp: tp, walTicker: walTicker}
+	writeAndWait := func() error {
+		tx, err := w.write(ctx)
+		if err != nil {
+			return err
+		}
+		db.Load().Wait(tx)
+		return nil
+	}
+
+	errg := &errgroup.Group{}
+	errg.SetLimit(32)
+	commandDistribution := make(map[command]int)
+	for i := 0; i < numCommands; i++ {
+		cmd := genCommand()
+		commandDistribution[cmd]++
+		switch cmd {
+		case insert:
+			errg.Go(func() error {
+				// This is a hack to ensure some randomized goroutine
+				// scheduling.
+				// TODO(asubiotto): Figure out if we still need this.
+				time.Sleep(1 * time.Millisecond)
+				if _, err := w.write(ctx); err != nil && !canIgnoreError(err) {
+					return fmt.Errorf("insert error: %s", err)
+				}
+				return nil
+			})
+		case compact:
+			errg.Go(func() error {
+				// TODO(asubiotto): Maybe we should offer a finer-grained way to
+				// trigger leveled compaction.
+				if err := tp.GetTable().EnsureCompaction(); err != nil && !errors.Is(err, io.EOF) {
+					return fmt.Errorf("compaction error: %w", err)
+				}
+				return nil
+			})
+		case rotate:
+			errg.Go(func() error {
+				table := tp.GetTable()
+				if err := table.RotateBlock(ctx, table.ActiveBlock(), false); err != nil {
+					return fmt.Errorf("rotate error: %s", err)
+				}
+				return nil
+			})
+		case snapshot:
+			errg.Go(func() error {
+				if err := db.Load().Snapshot(ctx); err != nil {
+					return fmt.Errorf("snapshot err: %w", err)
+				}
+				return nil
+			})
+		case restart:
+			// This is a hack to ensure some goroutines are scheduled before
+			// this restart.
+			// TODO(asubiotto): Figure out if we still need this.
+			time.Sleep(1 * time.Millisecond)
+			// Graceful shutdown.
+			require.NoError(t, c.Close())
```

Comment:
Would be nice to have an ungraceful shutdown as well

Reply:
Yep! Definitely want to add that.